### PR TITLE
ARC-611 backfill from date

### PR DESCRIPTION
**Pull request by joshkay10**
**What's in this PR?**
Added a new endpoint argument for the  `/resync` call.
~~`commitHistoryDepth`~~ `commitsFromDate` which allows you to pass the date to fetch commits (currently doubles for main and branch commits, instead of two seperate depth values)
Example 
{...`"commitsFromDate": 2022-07-26`} or with timezone
{...`"commitsFromDate": 2022-07-26T23:47:22.393Z'`}
It has higher importance than feature flag.

**Why**
If we want to run this from the command line we can!!
Also is groundwork so we can call it from the UI

**Runbook**
[HOW TO run backfill](https://hello.atlassian.net/wiki/spaces/PF/pages/1824654901/HOW+TO+run+backfill+resync+endpoint)

**Affected issues**  
ARC-611

**How has this been tested?**  
Updated the branch and commits unit tests in sync to make sure this value is used and used before ff value.

Tested in stage, tried install and `/resync` with combinations of values and feature flag settings.
Results as expected. 

**Whats Next?**
Create the UI which will need to pass a similar value

### CODE REVIEW

File: src/routes/api/api-router.ts
Comment (joshkay10) [inline src/routes/api/api-router.ts:86]: Went back and forth with commitHistoryDepth by itself or commitHistoryDepth + branchCommitHistoryDepth

Since there are two commit queries one for the main branch and one for other branches.

I ended up with one thats shared, because I was trying to limit the noise, but also the UI is only going to have 1 date picker(at this point 😂) so there was a consistency card to play there too....

File: src/sync/installation.ts
Comment (joshkay10) [inline src/sync/installation.ts:263]: This was the least evil way I could think of..... passing around the message payload, because of the cute way that this code exists all task processors being shared etc this was the most generic solution I found, basically just  goodies bag for everything now. 😂

File: src/routes/api/api-router.ts
Comment (gxueatlassian) [reply src/routes/api/api-router.ts:86]: /nit

I am okay with it not too fuzzy, but personally find it `xxxDepth` is a number like indicate much layer was nested inside a directory.

If the `commitHistoryDepth` means "xxx in milliseconds", I find it easier to understand if named `commitHistoryDepthInMilliseconds`

**Review by gxueatlassian — APPROVED** [review]
LGTM, but maybe wait for one more approval?

File: src/routes/api/api-router.ts
Comment (joshkay10) [reply src/routes/api/api-router.ts:86]: Fair, adding `milliseconds` os clear. naming is hard. lolol

The other option is align with the feature flag names `commitTimeLimit`

File: src/routes/api/api-router.ts
Comment (mboudreau) [inline src/routes/api/api-router.ts:86]: Why `|| undefined` here?

File: src/routes/api/api-router.ts
Comment (mboudreau) [inline src/routes/api/api-router.ts:93]: if you don't do `|| undefined` above, you can just do `if(!commitTimeLimit || commitTimeLimit <=0)`

File: src/routes/api/api-router.ts
Comment (mboudreau) [reply src/routes/api/api-router.ts:86]: ah, nvm, I can see why now as it's being passed.

File: src/routes/api/api-router.ts
Comment (mboudreau) [reply src/routes/api/api-router.ts:93]: actually, you can do the same even with `|| undefined` above

File: src/routes/api/api-router.ts
Comment (mboudreau) [inline src/routes/api/api-router.ts:101]: This I'm not sure I agree with, now that I think about it.  In the UI, it'll be something like "backfill from 7 days ago", which is fine, but I think that should be a date coming in specifically as it should be based on the user's current timezone and not the server's.

That means that the 'timeLimit' in this case is specifically a date which should be filtered down to the graphql query which is a GitTimestamp (a version of date).

File: src/sqs/backfill.ts
Comment (mboudreau) [inline src/sqs/backfill.ts:12]: This should be date

File: src/sync/installation.ts
Comment (mboudreau) [reply src/sync/installation.ts:265]: yeah, this is an acceptable solution.  This whole sync functionality needs to be redone...

File: src/sync/installation.ts
Comment (mboudreau) [inline src/sync/installation.ts:47]: Don't we have types for these message payloads?  Pretty sure the queues use a generic, but there's no reason we can't do the same here and let each processor have their own typings to make it a bit safer

**Review by mboudreau — CHANGES_REQUESTED** [review]
I think there's a core issue here that needs addressing but good first pass.  Happy to chat some more through it

File: src/routes/api/api-router.ts
Comment (joshkay10) [reply src/routes/api/api-router.ts]: change to be even even more explicit with commitsFromDate string

File: src/sync/installation.ts
Comment (joshkay10) [reply src/sync/installation.ts]: Used correct types

File: src/sync/branches.ts
Comment (gxueatlassian) [inline src/sync/branches.ts:8]: /nit
Is `~/src/blah` a pattern we prefer? Or team prefer `./blah` (or even `../../blah` if is another folder?)

File: src/sync/sync-utils.ts
Comment (gxueatlassian) [inline src/sync/sync-utils.ts:37]: /nit
`flagName` could be more strict typed as 
`SYNC_MAIN_COMMIT_TIME_LIMIT  | SYNC_BRANCH_COMMIT_TIME_LIMIT`

**Review by gxueatlassian — DISMISSED** [review]

File: src/sync/branches.ts
Comment (mboudreau) [reply src/sync/branches.ts:8]: if it's in the same folder, use `./blah`, if it's in a shortcut folder (like models) use `models/blah`, for anything else, use `~/src/some-dir/blah`

File: src/sync/sync-utils.ts
Comment (mboudreau) [inline src/sync/sync-utils.ts:11]: I'd made this a date, easier typing to deal with.

File: src/routes/api/api-router.ts
Comment (mboudreau) [inline src/routes/api/api-router.ts:93]: Why aren't you keeping the parsed date to pass it on to the next function?  Better typings than just a string :)

Comment (mboudreau) [reply src/routes/api/api-router.ts:93]: also, this should check to see that the date is valid (like, not be in the future with `date < new Date()`)

File: src/sqs/backfill.ts
Comment (mboudreau) [inline src/sqs/backfill.ts:12]: Date here

File: src/sync/sync-utils.ts
Comment (mboudreau) [reply src/sync/sync-utils.ts:37]: yeah, agreed, it should just be the flags we do care about for this, which is the 2 in `NumberFlags` but you don't want your typings to include others if another one is added in the future.

File: src/routes/api/api-router.ts
Comment (joshkay10) [reply src/routes/api/api-router.ts]: Have converted to Dates when appropriate 🤟

File: src/github/client/github-installation-client.ts
Comment (joshkay10) [inline src/github/client/github-installation-client.ts:221]: A date right until the string is required 😊

File: src/sync/sync-utils.ts
Comment (joshkay10) [reply src/sync/sync-utils.ts]: typed!

File: src/routes/api/api-router.ts
Comment (mboudreau) [inline src/routes/api/api-router.ts:88]: you're changing the type of the object here from unknown to date.  I'd probably just stick to Date to make sure there's no confusion.

**Review by mboudreau — APPROVED** [review]

**Review by gxueatlassian — APPROVED** [review]

**Pull request merged**
